### PR TITLE
Converted to walk function instead of a RuleWalker

### LINKED ISSUE #21: Convert to a walk function instead of a RuleWalker?
See discussion in https://github.com/palantir/tslint/issues/2522.

Tracking issue behind changes being split out of #20.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Fixes #21

### CODE REVIEW

File: noCircularImportsRule.ts [hunk 1]
```diff
@@ -1,8 +1,15 @@
 import { relative, sep } from 'path';
 import * as Lint from 'tslint';
-import { IOptions } from 'tslint/lib/language/rule/rule';
 import * as ts from 'typescript';
 
+interface Options {
+  /** @internal */
```
Comment (bcherny) [inline noCircularImportsRule.ts]: What's this for?

File: noCircularImportsRule.ts [hunk 2]
```diff
@@ -34,51 +48,44 @@ const imports = new Map<string, Map<string, ts.Node>>()
...
+      // Slice the array so we don't match this file twice.
+      if (maybeCycle.slice(1, -1).some(fileName => found.has(fileName))) {
+          continue
       }
+      maybeCycle.forEach(x => found.add(x))
+      const node = imports.get(fileName) !.get(maybeCycle[1]) !
+
+      context.addFailureAt(node.getStart(), node.getWidth(), Rule.FAILURE_STRING + ": " + maybeCycle
+          .concat(fileName)
+          .map(x => relative(context.options.rootDir, x))
```
Comment (bcherny) [inline noCircularImportsRule.ts]: Nice!

File: noCircularImportsRule.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply noCircularImportsRule.ts]: Yeah I was a little confused how to represent this. It's not common for the `interface Options` on a rule to be different from the config the rule is allowed to take in.

`@internal` is recognized by TS and seems to be coming up as part of [TSDoc](https://github.com/Microsoft/tsdoc/issues/8), so I figured it was appropriate here. If you change or remove this I won't be offended at all 😉

**Comment by bcherny** [comment]
Thanks @JoshuaKGoldberg!

**Pull request merged**